### PR TITLE
Rewrite edit view

### PR DESCRIPTION
**Pull request by dfaulken**
How does this seem as a way of setting up the edit page?

*Should merge* once approved, as the base is `rewrite`. Then the final step is the controller.

### CODE REVIEW

**Comment by dfaulken** [comment]
screen shot 2018-06-16 at 12 02 07 pm

File: app/models/bus_stop.rb [hunk 1]
```diff
@@ -1,155 +1,32 @@
...
-  include DateAndTimeMethods
   has_paper_trail
-  validates :name, presence: true
-  validates :hastus_id, presence: true, uniqueness: true
+
   has_and_belongs_to_many :routes
   has_many :bus_stop_fields
   has_many :fields, through: :bus_stop_fields
 
+  accepts_nested_attributes_for :bus_stop_fields
```
Comment (dfaulken) [inline app/models/bus_stop.rb]: This is the only real change in this section. Everything else I just moved around for clarity.

File: app/models/field.rb [hunk 2]
```diff
@@ -6,12 +6,18 @@ class Field < ApplicationRecord
...
+
+  scope :in_category, ->(category) { where category: category }
+
   def multiple_choice?
     field_type == :choice
   end
+
+  def self.categories
+    order(:category).pluck(:category).uniq
+  end
```
Comment (dfaulken) [inline app/models/field.rb]: These methods are only used from the bus stop model, but I still prefer keeping some of the field-related functionality in its own model file.

File: app/views/bus_stops/_form.haml [hunk 3]
```diff
@@ -6,37 +6,34 @@
...
+        %th
+          Stop Info
+    %tbody
       %tr
         %td= f.label :name
         %td= stop.name
       %tr
         %td= f.label :hastus_id, 'Hastus ID'
         %td= stop.hastus_id
-      - if current_user.admin?
```
Comment (dfaulken) [inline app/views/bus_stops/_form.haml]: I do not know why we would wish to only show the route list to admins. Potentially a holdover from when it was editable?

File: db/seeds.rb [hunk 4]
```diff
@@ -63,7 +46,7 @@
 stops.each do |route_number, stop_names|
   stop_names.each do |stop_name|
     # Anytime in the last two months
-    Timecop.freeze rand(86_400).minutes.ago do
+    Timecop.freeze Time.now - rand(2.months) do
```
Comment (dfaulken) [inline db/seeds.rb]: I saw this and had to wonder why I didn't just do it originally.

File: lib/tasks/create_bus_stop_fields.rake [hunk 5]
```diff
@@ -268,6 +268,7 @@ namespace :bus_stop_fields do
       ]
     }
 
+    Field.delete_all
```
Comment (dfaulken) [inline lib/tasks/create_bus_stop_fields.rake]: Just in case

Comment (mboneil10) [reply lib/tasks/create_bus_stop_fields.rake]: Good call

File: app/models/bus_stop.rb [hunk 6]
```diff
@@ -1,155 +1,32 @@
...
-    security_and_safety: SECURITY_AND_SAFETY,
-    technology: TECHNOLOGY
-  }.freeze
-
-  SUPER_HASH.each do |hash|
-    hash.each do |name, options|
-      if options.kind_of?(Array)
-        validates name, inclusion: { in: options }, allow_blank: true
+  def data_fields
+    Field.categories.map do |category|
```
Comment (mboneil10) [inline app/models/bus_stop.rb]: Cool, this whole method is good.

File: app/models/field.rb [hunk 7]
```diff
@@ -6,12 +6,18 @@ class Field < ApplicationRecord
...
   validates :category, presence: true
   validates :rank, presence: true, uniqueness: { scope: :category }
-  validates :field_type, inclusion: { in: %i[boolean choice] }
+  validates :field_type, inclusion: { in: %w[boolean choice] }
   
   serialize :choices, Array
   validates :choices, presence: true, if: :multiple_choice?
-  
+
+  scope :in_category, ->(category) { where category: category }
```
Comment (mboneil10) [inline app/models/field.rb]: 👍

File: app/models/field.rb (hunk 2 above)
Comment (mboneil10) [reply app/models/field.rb]: I agree with that preference.

File: app/views/bus_stops/_form.haml [hunk 8]
```diff
@@ -6,37 +6,34 @@
...
-              - else
-                = f.check_box name
+      - @data_fields.each_pair do |category, category_fields|
+        %tr
+          %th{ colspan: 2 }= category
+        - category_fields.each do |data_field|
+          = fields_for data_field do |data_field_form|
+            %tr
+              %td= data_field_form.label :value, data_field.field_name
+              - case data_field.field.field_type
```
Comment (mboneil10) [inline app/views/bus_stops/_form.haml]: Hooray!

File: db/seeds.rb [hunk 9]
```diff
@@ -12,24 +12,7 @@
...
-              field_type: :boolean,
-              description: 'Determine if this stop has solar lighting.',
-              rank: 1
-
-Field.create! name: 'Completed',
-              category: 'Info',
-              field_type: :boolean,
-              description: 'Have all attributes been checked?',
-              rank: 1
+# Recommended to create fields with `rails bus_stop_fields:create`.
```
Comment (mboneil10) [inline db/seeds.rb]: Ok

File: app/views/bus_stops/_form.haml (hunk 8 above)
Comment (dfaulken) [reply app/views/bus_stops/_form.haml]: Definitely very open to suggestions for better local variable names.

File: db/seeds.rb (hunk 9 above)
Comment (dfaulken) [reply db/seeds.rb]: Seems like we wouldn't want a different set of fields in dev vs. prod, similar to how right now we don't have a different set of columns on the bus stop table.

Comment (mboneil10) [reply db/seeds.rb]: Got i

File: app/views/bus_stops/_form.haml (hunk 8 above)
Comment (mboneil10) [reply app/views/bus_stops/_form.haml]: Too many `field`s for you?

File: app/controllers/bus_stops_controller.rb [hunk 10]
```diff
@@ -77,6 +77,7 @@ def update
 
   def find_stop
     @stop = BusStop.find_by hastus_id: params.require(:id)
+    @data_fields = @stop.data_fields
```
Comment (Anbranin) [inline app/controllers/bus_stops_controller.rb]: I wouldn't have bothered assigning this to its own instance variable personally.

Comment (Anbranin) [reply app/controllers/bus_stops_controller.rb]: Actually reading further I see why you did it.

File: app/models/bus_stop.rb [hunk 11]
```diff
@@ -172,16 +49,7 @@ def route_list
...
-              LIMITED_ATTRS
-            else Hash[columns.map { |c| [c.name, c.name.humanize] }]
-            end
-    CSV.generate headers: true do |csv|
-      csv << attrs.values
-      all.each do |stop|
-        csv << attrs.keys.map { |attr| stop.send attr }
-      end
-    end
+    # TODO
```
Comment (Anbranin) [inline app/models/bus_stop.rb]: This is OK?

File: app/controllers/bus_stops_controller.rb (hunk 10 above)
Comment (dfaulken) [reply app/controllers/bus_stops_controller.rb]: Collect your data in the controller, was my thinking.

File: app/models/bus_stop.rb (hunk 11 above)
Comment (dfaulken) [reply app/models/bus_stop.rb]: Out of scope for this PR, I think, but should definitely be done before we merge to master, agreed.

File: app/views/bus_stops/_form.haml (hunk 8 above)
Comment (Anbranin) [reply app/views/bus_stops/_form.haml]: One app where we have to make a form for a model called "Field" is too many

**Pull request merged**